pkg/wal/wal.go

```diff
@@ -42,19 +75,759 @@ type Segment interface {
+}
+
+func (log *log) flushBuffer(buffer buffer) error {
+	if buffer.count == 0 {
+		return nil
+	}
+
+	defer func() {
+		log.bytesBuffer.Reset()
+		log.timestampsBuffer.Reset()
+	}()
+
+	// placeholder, preset batch length value is 0
+	batchLen := 0
+	if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int64(batchLen)); err != nil {
+		return errors.Wrap(err, "Write batch length error")
+	}
+	for seriesID, timestamps := range buffer.timestampMap {
+		// Generate seriesID binary
+		seriesIDBytes := seriesID.Marshal()
+		seriesIDBytesLen := len(seriesIDBytes)
+
+		// Generate timestamps compression binary
+		log.timestampsBuffer.Reset()
+		timestampWriter := encoding.NewWriter()
+		timestampEncoder := encoding.NewXOREncoder(timestampWriter)
+		timestampWriter.Reset(log.timestampsBuffer)
+		for _, timestamp := range timestamps {
+			timestampEncoder.Write(timeTouUnixNano(timestamp))
+		}
+		timestampWriter.Flush()
+		timestampsBytes := log.timestampsBuffer.Bytes()
+		timestampsBytesLen := len(timestampsBytes)
+
+		// Generate values compression binary
+		valuesBytes := snappy.Encode(nil, buffer.valueMap[seriesID])
+
+		// Write entry data
+		entryLen := seriesIDLength + seriesIDBytesLen + seriesCountLength + timestampsBinaryLength + timestampsBytesLen + len(valuesBytes)
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int64(entryLen)); err != nil {
+			return errors.Wrap(err, "Write entry length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int16(seriesIDBytesLen)); err != nil {
+			return errors.Wrap(err, "Write seriesID length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, seriesIDBytes); err != nil {
+			return errors.Wrap(err, "Write seriesID error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int32(len(timestamps))); err != nil {
+			return errors.Wrap(err, "Write series count error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, int16(timestampsBytesLen)); err != nil {
+			return errors.Wrap(err, "Write timestamps length error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, timestampsBytes); err != nil {
+			return errors.Wrap(err, "Write timestamps error")
+		}
+		if err := binary.Write(log.bytesBuffer, binary.LittleEndian, valuesBytes); err != nil {
+			return errors.Wrap(err, "Write values error")
+		}
+	}
+	// Rewrite batch length
+	batchBytes := log.bytesBuffer.Bytes()
+	batchLen = len(batchBytes) - batchLength
+	rewriteInt64InBuf(batchBytes, int64(batchLen), 0, binary.LittleEndian)
+
+	return log.writeWorkSegment(batchBytes)
+}
+
+func (log *log) swapWorkSegment() (Segment, error) {
+	log.workSegmentMutex.Lock()
+	defer log.workSegmentMutex.Unlock()
+
+	if err := log.workSegment.file.Close(); err != nil {
+		return nil, errors.Wrap(err, "Close WAL segment error")
+	}
+
+	oldSegment := log.workSegment
+	// Create new segment.
+	segmentID := log.workSegment.segmentID + 1
```

Comment:
What will happen if the segmentID overflow? In the loading process, the max id will be the work id. Such an assumption will fail if data overflow happens.